File: Sources/AgentCoding/CloudUploader.swift
import Foundation
import Security
import SandboxEngine

/// Batches `BACCloudEvent`s and POSTs them to the analytics service's
/// `/ac-ingest` endpoint over mTLS. The install authenticates with the
/// leaf cert issued from the workspace's org CA — no bearer token rides
/// with the data.
///
/// In-memory only: dropping buffered events on a hard quit is fine for
/// telemetry, and a disk-backed retry queue is a Phase 3c-or-later
/// problem. Up to 200 events / 5 s in flight at once. Failures are
/// logged but don't drop the buffer — the next flush retries the whole
/// batch.
public final class BACCloudUploader: @unchecked Sendable {
    /// Events to flush. Guarded by `lock`.
    private var pending: [BACCloudEvent] = []
    private let lock = NSLock()

    /// Cap matching the server-side `MAX_EVENTS_PER_BATCH`. Bumping
    /// this requires bumping the server too.
    private let maxBatch = 500
    /// Auto-flush threshold — when pending hits this, kick a flush
    /// without waiting for the timer.
    private let flushHighWatermark = 200
    /// Periodic flush interval. Long enough that a Claude turn that
    /// fires 5 events bundles them into one POST; short enough that
    /// the admin UI feels live.
    private let flushIntervalSec: TimeInterval = 5

    private var flushTask: Task<Void, Never>?
    /// Set by `shutdown()` to stop the periodic flush loop. Reads
    /// happen under `lock`.
    private var stopped = false

    /// Endpoint to POST batches to. Defaults to `analytics.bromure.io`,
    /// overridable via `BROMURE_AC_INGEST_URL` for dev/staging.
    private let endpoint: URL
    /// URLSession configured with the mTLS challenge handler. Held by
    /// the uploader so the connection pool / TLS session cache survives
    /// across flushes — re-handshaking on every batch would defeat the
    /// point of keeping the session warm.
    private let session: URLSession
    private let delegate: BACMTLSDelegate

    public convenience init() {
        self.init(endpoint: BACEnrollment.defaultAnalyticsURL)
    }

    public init(endpoint: URL) {
        self.endpoint = endpoint
        self.delegate = BACMTLSDelegate()
        let config = URLSessionConfiguration.ephemeral
        config.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
        flushTask = Task { [weak self] in
            await self?.flushLoop()
        }
    }

    public func enqueue(_ event: BACCloudEvent) {
        var shouldKick = false
        var pendingAfter = 0
        var dropped = 0
        lock.lock()
        pending.append(event)
        if pending.count >= flushHighWatermark { shouldKick = true }
        // Hard cap: if a long network outage piles up more than a
        // couple of multiples of the server limit, drop the oldest
        // half. Keeping the buffer unbounded would eat memory; an
        // admin gets at-least-recent visibility either way.
        let drainCap = maxBatch * 4
        if pending.count > drainCap {
            dropped = pending.count - drainCap / 2
            pending.removeFirst(dropped)
        }
        pendingAfter = pending.count
        lock.unlock()
        BACDebug.log("[bac/uploader]",
                     "enqueue eventType=\(event.eventType) pending=\(pendingAfter) dropped=\(dropped) shouldKick=\(shouldKick)")
        if shouldKick { Task { await self.flushNow() } }
    }

    public func shutdown() {
        lock.lock()
        stopped = true
        lock.unlock()
        flushTask?.cancel()
        session.finishTasksAndInvalidate()
    }

    private func flushLoop() async {
        while !Task.isCancelled {
            try? await Task.sleep(nanoseconds: UInt64(flushIntervalSec * 1_000_000_000))
            if Task.isCancelled { break }
            await flushNow()
        }
    }

    public func flushNow() async {
        // Pull a slice off the front (matched by maxBatch) and try
        // sending it. Holding the lock across the network call would
        // serialise enqueue → keep the lock briefly, copy out, then
        // POST without it.
        let (batch, isStopped): ([BACCloudEvent], Bool) = lock.withLock {
            var b: [BACCloudEvent] = []
            if !pending.isEmpty {
                let take = min(pending.count, maxBatch)
                b = Array(pending.prefix(take))
            }
            return (b, stopped)
        }
        if isStopped {
            BACDebug.log("[bac/uploader]", "flushNow noop (stopped)")
            return
        }
        if batch.isEmpty {
            BACDebug.log("[bac/uploader]", "flushNow noop (empty)")
            return
        }

        // Confirm the install is still enrolled and that we have an
        // mTLS identity to present. Either being absent means the next
        // POST would 401/403; bail early so the buffer survives until
        // enrollment / cert-rotation is back.
        guard BACEnrollmentStore.load() != nil else {
            // Lost enrollment between enqueue and flush. Drop the
            // batch — a future enrollment will start fresh.
            BACDebug.log("[bac/uploader]",
                         "flushNow drop \(batch.count) events (not enrolled)")
            lock.withLock {
                pending.removeFirst(min(pending.count, batch.count))
            }
            return
        }

        let postT0 = Date()
        BACDebug.log("[bac/uploader]",
                     "POST start endpoint=\(endpoint.absoluteString) events=\(batch.count)")
        do {
            try await postBatch(batch)
            BACDebug.log("[bac/uploader]",
                         "POST done events=\(batch.count) took=\(BACDebug.ms(postT0))")
            // Success: drop the prefix that matched what we sent.
            lock.withLock {
                pending.removeFirst(min(pending.count, batch.count))
            }
        } catch {
            BACDebug.log("[bac/uploader]",
                         "POST failed events=\(batch.count) took=\(BACDebug.ms(postT0)) error=\(error)")
            // Leave `pending` intact for the next flush. Log so an
            // operator can spot a sustained failure (expired leaf
            // cert, missing org CA, etc.) without a disk-backed retry
            // queue.
            FileHandle.standardError.write(Data(
                "[bac/uploader] flush of \(batch.count) events failed: \(error)\n".utf8))
        }
    }

    private func postBatch(_ events: [BACCloudEvent]) async throws {
        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let body = try encoder.encode(["events": events])
        req.httpBody = body
        BACDebug.log("[bac/uploader]",
                     "URLSession.data start bodyBytes=\(body.count)")
        let (_, resp) = try await session.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? -1
        BACDebug.log("[bac/uploader]",
                     "URLSession.data done status=\(status)")
        guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw NSError(domain: "BAC.Uploader", code: status,
                          userInfo: [NSLocalizedDescriptionKey: "HTTP \(status)"])
        }
    }
}

/// URLSession delegate that answers TLS client-cert challenges with the
/// install's leaf cert via `BACMTLSIdentity`. Server trust is left to
/// the system default — the analytics service uses an ACME-issued
/// public cert, so the regular root-store path applies.
private final class BACMTLSDelegate: NSObject, URLSessionDelegate {
    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void,
    ) {
        let method = challenge.protectionSpace.authenticationMethod
        if method == NSURLAuthenticationMethodClientCertificate {
            do {
                let identity = try BACMTLSIdentity.current()
                let cred = URLCredential(identity: identity, certificates: nil, persistence: .forSession)
                completionHandler(.useCredential, cred)
            } catch {
                FileHandle.standardError.write(Data(
                    "[bac/uploader] no mTLS identity: \(error)\n".utf8))
                completionHandler(.cancelAuthenticationChallenge, nil)
            }
            return
        }
        completionHandler(.performDefaultHandling, nil)
    }
}

File: windows/Bromure.Cloud/CloudUploader.cs
using System.Net.Http;
using System.Net.Http.Json;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Bromure.Cloud;

/// <summary>
/// Direct port of <c>Sources/AgentCoding/CloudUploader.swift</c>.
///
/// <para>Batches <see cref="CloudEvent"/>s and POSTs them to the
/// analytics service's <c>/ac-ingest</c> endpoint over mTLS. The install
/// authenticates with the leaf cert issued from the workspace's org CA
/// — no bearer token rides with the data.</para>
///
/// <para>In-memory only: dropping buffered events on a hard quit is fine
/// for telemetry, and a disk-backed retry queue is a Phase 3c-or-later
/// problem. Up to 200 events / 5 s in flight at once. Failures are
/// logged but don't drop the buffer — the next flush retries the whole
/// batch.</para>
/// </summary>
public sealed class CloudUploader : IAsyncDisposable
{
    /// Cap matching the server-side <c>MAX_EVENTS_PER_BATCH</c>.
    /// Bumping this requires bumping the server too.
    private const int MaxBatch = 500;
    /// Auto-flush threshold — when pending hits this, kick a flush
    /// without waiting for the timer.
    private const int FlushHighWatermark = 200;
    /// Periodic flush interval. Long enough that a Claude turn that
    /// fires 5 events bundles them into one POST; short enough that
    /// the admin UI feels live.
    private static readonly TimeSpan FlushInterval = TimeSpan.FromSeconds(5);

    private readonly Uri _endpoint;
    private readonly HttpClient _client;
    private readonly ILogger _log;
    private readonly object _gate = new();
    private readonly List<CloudEvent> _pending = new();
    private CancellationTokenSource? _cts;
    private Task? _flushLoop;
    private bool _stopped;

    public CloudUploader(Uri endpoint, ICloudMtlsIdentity mtls, ILogger? log = null)
    {
        _endpoint = endpoint;
        _log = log ?? NullLogger.Instance;

        // Hook the SocketsHttpHandler's TLS callback to present the
        // mTLS identity when the server requests a client cert.
        var handler = new SocketsHttpHandler
        {
            SslOptions = new SslClientAuthenticationOptions
            {
                ClientCertificates = mtls.ClientCertificates,
                LocalCertificateSelectionCallback = (_, _, _, _, _) => mtls.SelectCertificate()!,
            },
        };
        _client = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(30) };

        _cts = new CancellationTokenSource();
        _flushLoop = Task.Run(() => FlushLoopAsync(_cts.Token));
    }

    public void Enqueue(CloudEvent ev)
    {
        bool kick;
        int dropped = 0;
        lock (_gate)
        {
            _pending.Add(ev);
            kick = _pending.Count >= FlushHighWatermark;
            // Hard cap at 4× MaxBatch — drop the oldest half on a long outage.
            var drainCap = MaxBatch * 4;
            if (_pending.Count > drainCap)
            {
                dropped = _pending.Count - drainCap / 2;
                _pending.RemoveRange(0, dropped);
            }
        }
        if (dropped > 0) _log.LogWarning("[cloud] dropped {N} events (over cap)", dropped);
        if (kick) _ = Task.Run(FlushNowAsync);
    }

    public async Task FlushNowAsync()
    {
        List<CloudEvent> batch;
        bool stopped;
        lock (_gate)
        {
            stopped = _stopped;
            batch = _pending.Count == 0
                ? new List<CloudEvent>()
                : _pending.GetRange(0, Math.Min(_pending.Count, MaxBatch));
        }
        if (stopped || batch.Count == 0) return;

        try
        {
            await PostBatchAsync(batch).ConfigureAwait(false);
            lock (_gate)
            {
                _pending.RemoveRange(0, Math.Min(_pending.Count, batch.Count));
            }
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "[cloud] flush of {N} events failed", batch.Count);
            // Leave _pending intact for the next flush.
        }
    }

    private async Task PostBatchAsync(List<CloudEvent> events)
    {
        using var req = new HttpRequestMessage(HttpMethod.Post, _endpoint);
        var bodyJson = JsonSerializer.Serialize(new { events });
        req.Content = new StringContent(bodyJson, System.Text.Encoding.UTF8, "application/json");
        using var resp = await _client.SendAsync(req).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw new HttpRequestException($"HTTP {(int)resp.StatusCode}");
        }
    }

    private async Task FlushLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await Task.Delay(FlushInterval, ct).ConfigureAwait(false); }
            catch (OperationCanceledException) { return; }
            await FlushNowAsync().ConfigureAwait(false);
        }
    }

    public async ValueTask DisposeAsync()
    {
        lock (_gate) _stopped = true;
        try { _cts?.Cancel(); } catch { }
        if (_flushLoop is not null)
        {
            try { await _flushLoop.ConfigureAwait(false); } catch { }
        }
        _cts?.Dispose();
        _client.Dispose();
    }
}

/// <summary>
/// Provides the install's mTLS leaf cert + the selection callback for
/// SocketsHttpHandler. Direct port of the macOS <c>BACMTLSIdentity</c>
/// abstraction.
/// </summary>
public interface ICloudMtlsIdentity
{
    X509CertificateCollection ClientCertificates { get; }
    X509Certificate? SelectCertificate();
}
